### PR TITLE
optimise arrayOps in strawman collection

### PR DESCRIPTION
**Pull request by ackratos**
Benchmarked array slice performance, @mkeskells optimization have performance gain to scala.ArrayOps implementation and a big win of strawman collection. I am trying to optimize strawman implementation (but not sure whether its a good idea considered current implementation has a good, unified abstraction on View)

```
ArrayBaselineBenchmark
[info] Benchmark                               (size)  (vLoSize)  Mode  Cnt         Score         Error  Units
[info] ArrayBaselineBenchmark.sliceLaterHalf       39         39  avgt   30        66.163 ?      14.964  ns/op
[info] ArrayBaselineBenchmark.sliceLaterHalf      282         39  avgt   30       701.774 ?     320.821  ns/op
[info] ArrayBaselineBenchmark.sliceLaterHalf    73121         39  avgt   30    115300.047 ?   40758.378  ns/op
[info] ArrayBaselineBenchmark.sliceLaterHalf  7312102         39  avgt   30  15430154.435 ? 3136356.950  ns/op

ArrayView implementation
[info] Benchmark                           (size)  (vLoSize)  Mode  Cnt         Score         Error  Units
[info] ArrayViewBenchmark.sliceLaterHalf       39         39  avgt   30       767.224 ?     257.456  ns/op
[info] ArrayViewBenchmark.sliceLaterHalf      282         39  avgt   30      5511.769 ?    2177.406  ns/op
[info] ArrayViewBenchmark.sliceLaterHalf    73121         39  avgt   30   1000808.203 ?   12135.514  ns/op
[info] ArrayViewBenchmark.sliceLaterHalf  7312102         39  avgt   30  86467759.270 ? 1719425.377  ns/op

Mike https://github.com/scala/scala/pull/5652
[info] Benchmark                               (size)  (vLoSize)  Mode  Cnt         Score         Error  Units
[info] ArrayBaselineBenchmark.sliceLaterHalf       39         39  avgt   30        33.081 ?       0.700  ns/op
[info] ArrayBaselineBenchmark.sliceLaterHalf      282         39  avgt   30       221.226 ?       4.331  ns/op
[info] ArrayBaselineBenchmark.sliceLaterHalf    73121         39  avgt   30     67573.426 ?    9092.555  ns/op
[info] ArrayBaselineBenchmark.sliceLaterHalf  7312102         39  avgt   30  17210746.261 ? 4301349.666  ns/op
```

### CODE REVIEW

**Comment by julienrf** [comment]
Hey, thanks for taking this initiative but I’m afraid this is going to overlap with #278.

File: collections/src/main/scala/strawman/collection/package.scala [hunk 1]
```diff
@@ -12,7 +12,20 @@ package object collection extends LowPriority {
   implicit def stringToStringOps(s: String): immutable.StringOps = new immutable.StringOps(s)
 
   /** Decorator to add collection operations to arrays. */
-  implicit def arrayToArrayOps[A](as: Array[A]): ArrayOps[A] = new ArrayOps[A](as)
+  object ArrayOpsDecorators {
```
Comment (szeiger) [inline collections/src/main/scala/strawman/collection/package.scala]: Needs a fallback for `Array[Any]` that dispatches dynamically

**Comment by ackratos** [comment]
@julienrf thanks for reminding, thankfully I have a chance to learn from #278

**Comment by ackratos** [comment]
@szeiger yes, I will regard #278 as my baseline. Thanks for the changes

**Comment by julienrf** [comment]
Hey @ackratos, now that #278 has been merged you can continue this work. I think this optimization is useful. I’m wondering what’s the cost of its implementation in terms of bytecode size.
Also, I noticed that `slice` could be optimized in general for indexed sequence (simply by implementing it in terms of `view.take(…).drop(…)` instead of `View.Drop(View.Take(…), …)`). Would you be interested in finishing this PR and also adding the optimization I just mentioned and then benchmarking everything?

**Comment by ackratos** [comment]
collection-strawman's Array slice still has space to enhance,  in following result, above is collection-strawman's result, below is Mike's optimization:
https://github.com/scala/scala/pull/5652
https://github.com/scala/scala/compare/2.12.x...rorygraves:cong/2.12.x_ArrayTest?expand=1

```
[info] Result "access_slice":
[info]   3659363.534 ?(99.9%) 1674784.945 ns/op [Average]
[info]   (min, avg, max) = (2646958.595, 3659363.534, 5549813.105), stdev = 875944.910
[info]   CI (99.9%): [1984578.589, 5334148.480] (assumes normal distribution)
[info] 
[info] 
[info] # Run complete. Total time: 00:04:58
[info] 
[info] Benchmark                              (size)  Mode  Cnt        Score         Error  Units
[info] ImmutableArrayBenchmark.access_slice        0  avgt    8       41.986 ?       2.875  ns/op
[info] ImmutableArrayBenchmark.access_slice        1  avgt    8      139.056 ?      34.593  ns/op
[info] ImmutableArrayBenchmark.access_slice        2  avgt    8       75.239 ?      51.465  ns/op
[info] ImmutableArrayBenchmark.access_slice        3  avgt    8      102.300 ?     100.948  ns/op
[info] ImmutableArrayBenchmark.access_slice        4  avgt    8      109.680 ?      58.196  ns/op
[info] ImmutableArrayBenchmark.access_slice        7  avgt    8       41.041 ?       2.172  ns/op
[info] ImmutableArrayBenchmark.access_slice        8  avgt    8       45.605 ?      12.347  ns/op
[info] ImmutableArrayBenchmark.access_slice       15  avgt    8      243.585 ?     386.930  ns/op
[info] ImmutableArrayBenchmark.access_slice       16  avgt    8       67.775 ?      32.322  ns/op
[info] ImmutableArrayBenchmark.access_slice       17  avgt    8       62.008 ?      26.068  ns/op
[info] ImmutableArrayBenchmark.access_slice       39  avgt    8      131.026 ?     164.174  ns/op
[info] ImmutableArrayBenchmark.access_slice      282  avgt    8      135.909 ?      48.593  ns/op
[info] ImmutableArrayBenchmark.access_slice     4096  avgt    8     1841.219 ?     867.728  ns/op
[info] ImmutableArrayBenchmark.access_slice   131070  avgt    8    55579.599 ?   21597.217  ns/op
[info] ImmutableArrayBenchmark.access_slice  7312102  avgt    8  3659363.534 ? 1674784.945  ns/op
[info] 
[info] Benchmark result is saved to /Users/zhaocong/Developer/collection-strawman/benchmarks/time/target/scala-2.13.0-M2/jmh-result.json
[success] Total time: 315 s, completed Jan 15, 2018 11:08:36 PM


[info] Result "scala.collection.mutable.ArrayBaselineBenchmark.access_slice":
[info]   1307237.310 ?(99.9%) 478467.896 ns/op [Average]
[info]   (min, avg, max) = (1020158.565, 1307237.310, 1817316.572), stdev = 250247.961
[info]   CI (99.9%): [828769.414, 1785705.207] (assumes normal distribution)
[info] 
[info] 
[info] # Run complete. Total time: 00:04:44
[info] 
[info] Benchmark                             (size)  Mode  Cnt        Score        Error  Units
[info] ArrayBaselineBenchmark.access_slice        0  avgt    8       14.009 ?      7.245  ns/op
[info] ArrayBaselineBenchmark.access_slice        1  avgt    8       18.760 ?     10.962  ns/op
[info] ArrayBaselineBenchmark.access_slice        2  avgt    8       14.603 ?      4.859  ns/op
[info] ArrayBaselineBenchmark.access_slice        3  avgt    8       12.027 ?      5.900  ns/op
[info] ArrayBaselineBenchmark.access_slice        4  avgt    8       18.682 ?      6.070  ns/op
[info] ArrayBaselineBenchmark.access_slice        7  avgt    8       26.929 ?     26.067  ns/op
[info] ArrayBaselineBenchmark.access_slice        8  avgt    8       30.137 ?     26.866  ns/op
[info] ArrayBaselineBenchmark.access_slice       15  avgt    8       40.062 ?     36.232  ns/op
[info] ArrayBaselineBenchmark.access_slice       16  avgt    8       42.642 ?     34.040  ns/op
[info] ArrayBaselineBenchmark.access_slice       17  avgt    8       19.694 ?      1.995  ns/op
[info] ArrayBaselineBenchmark.access_slice       39  avgt    8       21.482 ?      5.269  ns/op
[info] ArrayBaselineBenchmark.access_slice      282  avgt    8       93.949 ?    120.360  ns/op
[info] ArrayBaselineBenchmark.access_slice     4096  avgt    8      468.349 ?    197.093  ns/op
[info] ArrayBaselineBenchmark.access_slice   131070  avgt    8    11484.873 ?   3698.272  ns/op
[info] ArrayBaselineBenchmark.access_slice  7312102  avgt    8  1307237.310 ? 478467.896  ns/op
[success] Total time: 305 s, completed Jan 15, 2018 11:24:13 PM
```

**Comment by ackratos** [comment]
Close this as I have a new one contains less changes and easier to review: https://github.com/scala/collection-strawman/pull/354

**Pull request closed without merging**